src/storage/src/source/persist_source.rs

```diff
@@ -146,13 +166,27 @@ where
             .await
             .expect("cannot serve requested as_of");
 
-        loop {
+        let mut done = false;
+        while !done {
             while let Ok(leased_part) = consumed_part_rx.try_recv() {
                 subscription.return_leased_part(leased_part.into());
             }
 
-            yield subscription.next().await;
+            let (parts, progress) = subscription.next().await;
+            if timely::PartialOrder::less_equal(&until_clone, &progress) {
+                // Cease yielding parts after this batch.
+                done = true;
+            }
+            yield (parts, progress);
         }
+
+        // Rather than simply end, we spawn a task that can continue to return leases.
+        // This keeps the `ReadHandle` alive until we have confirmed each lease as read.
+        mz_ore::task::spawn(|| "LeaseReturner", async move {
```

Comment:
Ha! This solution is actually quite a bit simpler in code than I had guessed before looking at it.